dashboard/app/controllers/registrations_controller.rb

```diff
@@ -14,19 +14,50 @@ class RegistrationsController < Devise::RegistrationsController
     :migrate_to_multi_auth, :demigrate_from_multi_auth
   ]
   skip_before_action :verify_authenticity_token, only: [:set_student_information]
-  skip_before_action :clear_sign_up_session_vars, only: [:new, :begin_sign_up, :cancel, :create]
+  skip_before_action :clear_sign_up_session_vars, only: [:new, :begin_sign_up, :begin_creating_user, :cancel, :create]
 
   #
   # GET /users/sign_up
   #
   def new
     session[:user_return_to] ||= params[:user_return_to]
     if PartialRegistration.in_progress?(session)
-      user_params = params[:user] || ActionController::Parameters.new
-      user_params[:user_type] ||= session[:default_sign_up_user_type]
-      user_params[:email] ||= params[:email]
+      if params[:new_sign_up].present?
+        user_params = params || ActionController::Parameters.new
+        user_params[:user_type] = params[:user_type] ? params[:user_type] : session[:default_sign_up_user_type]
+        user_params[:email] = params[:email]
+        user_params[:age] = user_params[:user_type] == 'teacher' ? '21+' : params[:age]
+
+        # Set email and data transfer preferences
+        if user_params[:user_type] == 'teacher'
+          user_params[:email_preference_opt_in_required] = true
+          user_params[:email_preference_opt_in] = params[:email_preference_opt_in] ? 'yes' : 'no'
+          user_params[:email_preference_request_ip] = request.ip
+          user_params[:email_preference_source] = EmailPreference::ACCOUNT_SIGN_UP
+          user_params[:email_preference_form_kind] = '0'
+        elsif user_params[:user_type] == 'student'
+          user_params[:parent_email_preference_request_ip] = request.ip
+          user_params[:parent_email_preference_source] = EmailPreference::ACCOUNT_SIGN_UP
+        end
+
+        user_params[:data_transfer_agreement_accepted] = params[:data_transfer_agreement_accepted] == '1'
+        if user_params[:data_transfer_agreement_required] && params[:data_transfer_agreement_accepted]
+          user_params[:data_transfer_agreement_accepted] = true
+          user_params[:data_transfer_agreement_request_ip] = request.ip
+          user_params[:data_transfer_agreement_source] = User::ACCOUNT_SIGN_UP
+          user_params[:data_transfer_agreement_kind] = '0'
+          user_params[:data_transfer_agreement_at] = DateTime.now
```

Comment:
After consulting with @artem-vavilov it could be even better to do a separate [service object](https://medium.com/selleo/essential-rubyonrails-patterns-part-1-service-objects-1af9f9573ca1) to encapsulate the business logic from the `if PartialRegistration.in_progress?(session)` block here. I think it's a good opportunity to leave the registrations controller better than we found it. Artem created a [Services::Base](https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/lib/services/base.rb) class that also lets you call `:call` directly instead of having to call `new` each time.

Reply:
I recommend creating two distinct Service Objects:
- `@user = Services::PartialRegistration::NewUserBuilder.call(request)` for this part:
https://github.com/code-dot-org/code-dot-org/blob/439f03b6e555f28edbdc067d18578bf8877b4635/dashboard/app/controllers/registrations_controller.rb#L26-L54
- `@user = Services::PartialRegistration::UserBuilder.call(request)` for this part:
https://github.com/code-dot-org/code-dot-org/blob/439f03b6e555f28edbdc067d18578bf8877b4635/dashboard/app/controllers/registrations_controller.rb#L56-L59

Reply:
Since we intend to only keep one set of logic after we complete the A/B test of the old signup flow vs. new signup flow, do you think it would make sense to only move the new one to a service object?

Reply:
Certainly!